PINCache/PINOperationQueue.m

```diff
@@ -127,27 +135,79 @@ + (instancetype)sharedOperationQueue
+  return [self addOperation:block withPriority:PINOperationQueuePriorityDefault];
 }
 
 - (id <PINOperationReference>)addOperation:(dispatch_block_t)block withPriority:(PINOperationQueuePriority)priority
 {
-  id <PINOperationReference> reference = [self nextOperationReference];
+  PINOperation *operation = [PINOperation operationWithBlock:^(id data) { block(); }
+                                                   reference:[self nextOperationReference]
+                                                    priority:priority
+                                                  identifier:nil
+                                                        data:nil
+                                                  completion:nil];
+  [self lock];
+    [self locked_addOperation:operation];
+  [self unlock];
   
-  NSMutableOrderedSet *queue = [self operationQueueWithPriority:priority];
+  [self scheduleNextOperations:NO];
   
-  PINOperation *operation = [PINOperation operationWithBlock:block reference:reference priority:priority];
+  return operation.reference;
+}
+
+- (id<PINOperationReference>)addOperation:(PINOperationBlock)block withPriority:(PINOperationQueuePriority)priority identifier:(NSString *)identifier data:(id)data dataCoallescingBlock:(PINOperationDataCoallescingBlock)dataCoallescingBlock completion:(dispatch_block_t)completion
+{
+  PINOperation *operation = nil;
+  BOOL isNewOperation = NO;
   
   [self lock];
-    dispatch_group_enter(_group);
-    [queue addObject:operation];
-    [_queuedOperations addObject:operation];
-    [_referenceToOperations setObject:operation forKey:reference];
+    if (identifier != nil && (operation = [_identifierToOperations objectForKey:identifier]) != nil) {
+      // There is an exisiting operation with the provided identifier, let's coallescing these operations
+      if (dataCoallescingBlock != nil) {
+        operation.data = dataCoallescingBlock(operation.data, data);
+      }
+      
+      if (completion != nil) {
+        dispatch_block_t initialCompletion = operation.completion;
+        if (initialCompletion == nil) {
+          operation.completion = completion;
+        } else {
+          operation.completion = ^{
+            initialCompletion();
+            completion();
+          };
+        }
+      }
+    } else {
+      isNewOperation = YES;
+      operation = [PINOperation operationWithBlock:block
+                                         reference:[self nextOperationReference]
+                                          priority:priority
+                                        identifier:identifier
+                                              data:data
+                                        completion:completion];
+      [self locked_addOperation:operation];
+    }
   [self unlock];
   
-  [self scheduleNextOperations:NO];
+  if (isNewOperation) {
+    [self scheduleNextOperations:NO];
+  }
   
-  return reference;
+  return operation.reference;
```

Comment:
Is it safe to access operation.reference without holding the lock? Should we grab a copy before unlocking? I guess since we don't mutate it anywhere it's safe, but we should add a comment to that effect at the very least?